### PR TITLE
Rest Client Reactive: SmallRye Stork integration

### PR DESCRIPTION
**Pull request by michalszynkiewicz**

### CODE REVIEW

File: extensions/resteasy-reactive/rest-client-reactive/runtime/pom.xml [hunk 1]
```diff
@@ -18,6 +18,10 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-jaxrs-client-reactive</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.smallrye.stork</groupId>
+            <artifactId>smallrye-stork-microprofile</artifactId>
```
Comment (cescoffier) [inline extensions/resteasy-reactive/rest-client-reactive/runtime/pom.xml:23]: why stork-microprofile?

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java [hunk 2]
```diff
@@ -151,26 +164,80 @@ public void handle(Throwable failure) {
...
-        requestOptions.setMethod(HttpMethod.valueOf(state.getHttpMethod()));
-        requestOptions.setURI(uri.getPath() + (uri.getQuery() == null ? "" : "?" + uri.getQuery()));
-        requestOptions.setFollowRedirects(followRedirects);
-        requestOptions.setSsl(isHttps);
         Object readTimeout = state.getConfiguration().getProperty(QuarkusRestClientProperties.READ_TIMEOUT);
-        if (readTimeout instanceof Long) {
-            requestOptions.setTimeout((Long) readTimeout);
+        Uni<RequestOptions> requestOptions;
+        if (uri.getScheme().startsWith("stork")) {
+            boolean isHttps = "storks".equals(uri.getScheme());
```
Comment (cescoffier) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java:195]: so the plural of stork is secure :-D

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties [hunk 3]
```diff
@@ -0,0 +1,5 @@
+stork.hello-service.service-discovery=static
+stork.hello-service.service-discovery.1=${quarkus.http.host}:${quarkus.http.test-port}
```
[the remark this answers is not in the document]
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties:2]: the numbers don't really matter

File: extensions/resteasy-reactive/rest-client-reactive/runtime/pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/runtime/pom.xml:23]: microprofile config support for Stork. In the future, it may also bring in some CDI support or sth

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+stork.hello-service.load-balancer=stat-based
+
+hello2/mp-rest/url=stork://hello-service/hello
```
[the remark this answers is not in the document]
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties:5]: Failure: https://github.com/quarkusio/quarkus/pull/19998/files#diff-ea922f111f94ae50d76e1f78200508c9bc63f2315d879b35a229be421922280cR46

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties (hunk 3 above)
[the remark this answers is not in the document]
Comment (aureamunoz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties:2]: I think the `1` is the service id.

**Review by aureamunoz — APPROVED** [review]
LGTM

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties (hunk 3 above)
[the remark this answers is not in the document]
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties:2]: no, it was just my way to pass a list via properties :D

**Comment by geoand** [comment]
I'll try and take a look tomorrow.

Has it been tested in native?
How about dev-mode?

**Comment by michalszynkiewicz** [comment]
> I'll try and take a look tomorrow.
> 
> Has it been tested in native?
> How about dev-mode?

No, I'll add a native test, and if I can a dev mode test too

**Comment by geoand** [comment]
:+1

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java (hunk 2 above)
Comment (geoand) [reply independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java:195]: LOLOLOL

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java [hunk 5]
```diff
@@ -151,26 +164,80 @@ public void handle(Throwable failure) {
...
-        requestOptions.setPort(port);
-        requestOptions.setMethod(HttpMethod.valueOf(state.getHttpMethod()));
-        requestOptions.setURI(uri.getPath() + (uri.getQuery() == null ? "" : "?" + uri.getQuery()));
-        requestOptions.setFollowRedirects(followRedirects);
-        requestOptions.setSsl(isHttps);
         Object readTimeout = state.getConfiguration().getProperty(QuarkusRestClientProperties.READ_TIMEOUT);
-        if (readTimeout instanceof Long) {
-            requestOptions.setTimeout((Long) readTimeout);
+        Uni<RequestOptions> requestOptions;
+        if (uri.getScheme().startsWith("stork")) {
```
Comment (geoand) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java:194]: Is there no constant somewhere in the Stock API we could use?

Comment (michalszynkiewicz) [reply independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java:194]: I don't think there is but we sure can add one :)

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java [hunk 6]
```diff
@@ -151,26 +164,80 @@ public void handle(Throwable failure) {
...
+        return requestOptions.onItem()
+                .transform(r -> r.setMethod(HttpMethod.valueOf(state.getHttpMethod()))
+                        .setURI(uri.getPath() + (uri.getQuery() == null ? "" : "?" + uri.getQuery()))
+                        .setFollowRedirects(followRedirects))
+                .onItem().invoke(r -> {
+                    if (readTimeout instanceof Long) {
+                        r.setTimeout((Long) readTimeout);
+                    }
+                })
+                .flatMap(options -> AsyncResultUni.toUni(handler -> httpClient.request(options, handler)));
```
Comment (geoand) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java:232]: `flatMap`? I thought one of the points of `Mutiny` was to use proper names and not names that are legacy from functional programming :) Mutiny has better alternatives that this, no?

**Review by geoand — COMMENTED** [review]
Some mention of Stork will also need to appear in the reference documentation that you need to write for the REST Client :)

**Comment by michalszynkiewicz** [comment]
> Some mention of Stork will also need to appear in the reference documentation that you need to write for the REST Client :)

sure thing

File: bom/application/pom.xml [hunk 7]
```diff
@@ -3533,6 +3534,21 @@
...
+            </dependency>
+            <dependency>
+                <groupId>io.smallrye.stork</groupId>
+                <artifactId>smallrye-stork-service-discovery-static-list</artifactId>
+                <version>${smallrye-stork.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>io.smallrye.stork</groupId>
+                <artifactId>smallrye-stork-microprofile</artifactId>
+                <version>${smallrye-stork.version}</version>
```
[the remark this answers is not in the document]
Comment (michalszynkiewicz) [reply bom/application/pom.xml:3550]: I created an issue for it, we can discuss it there: https://github.com/smallrye/smallrye-stork/issues/68

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java (hunk 5 above)
Comment (michalszynkiewicz) [reply independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java:197]: done

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java (hunk 6 above)
Comment (michalszynkiewicz) [reply independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java]: changed to `onItem().transformToUni(...)`

I'm trying to use Mutiny methods everywhere but the well known and established functional method names are much more familiar to me and they sometimes sneak in :)

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties [hunk 8]
```diff
@@ -0,0 +1,5 @@
+stork.hello-service.service-discovery=static
+stork.hello-service.service-discovery.1=${quarkus.http.host}:${quarkus.http.test-port}
+stork.hello-service.load-balancer=stat-based
```
[the remark this answers is not in the document]
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties]: fixed

File: integration-tests/rest-client-reactive-stork/src/test/java/io/quarkus/it/rest/reactive/stork/RestClientReactiveStorkTest.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.response.Response;
+
+@QuarkusTest
+public class RestClientReactiveStorkTest {
+
+    static final String SLOW_RESPONSE = "hello, I'm a slow server";
+    static final String FAST_RESPONSE = "hello, I'm a fast server";
+    static WireMockServer slowServer;
```
Comment (geoand) [inline integration-tests/rest-client-reactive-stork/src/test/java/io/quarkus/it/rest/reactive/stork/RestClientReactiveStorkTest.java:26]: Would it not be preferable to use what is described in https://quarkus.io/guides/rest-client#using-a-mock-http-server-for-tests?

The reason is that the linked way starts Wiremocks before Quarkus and can affect the build via properties

Of course we would need 2 since you are starting 2 wiremocks

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java [hunk 10]
```diff
@@ -92,6 +102,27 @@ void announceFeature(BuildProducer<FeatureBuildItem> features) {
...
+            // find out all the implementation classes listed in the service files
+            Set<String> implementations;
+            try {
+                implementations = ServiceUtil.classNamesNamedIn(Thread.currentThread().getContextClassLoader(),
+                        service);
+            } catch (IOException e) {
+                throw new IllegalStateException("Failed to collect implementations of " + service, e);
+            }
+
+            services.produce(new ServiceProviderBuildItem(providerClass.getName(), implementations));
```
Comment (geoand) [inline extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:122]: Can't we use `ServiceProviderBuildItem#allProviders`?

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java [hunk 11]
```diff
@@ -129,13 +160,18 @@ void registerRestClientListenerForTracing(
     }
 
     @BuildStep
-    @Record(ExecutionTime.STATIC_INIT)
+    @Record(ExecutionTime.RUNTIME_INIT)
```
Comment (geoand) [inline extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:163]: Why this change? Can't we just have stork in its own build step that will use the runtime init phase?

File: extensions/resteasy-reactive/rest-client-reactive/runtime/pom.xml (hunk 1 above)
Comment (cescoffier) [reply extensions/resteasy-reactive/rest-client-reactive/runtime/pom.xml:23]: I would state `config` then.

**Comment by cescoffier** [comment]
Start to looks good. I wonder what we can handle at build time (see @geoand's comment). I need to think about the configuration, as configuring both stock and the rest-client looks a bit "extra-step-required".

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties (hunk 4 above)
[the remark this answers is not in the document]
Comment (cescoffier) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/resources/stork-application.properties:5]: We need to think about the proper way to configure things. We may need something simpler.

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java (hunk 11 above)
Comment (cescoffier) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:163]: I wonder how much we can do at build time. At least having a Stock build item would be good.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java (hunk 10 above)
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:122]: `allProviders` focuses on a single file. But I see there's also `allProvidersFromClassPath` which does exactly what I need here :)

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java (hunk 11 above)
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:163]: good catch, I need only Stork initialization moved to runtime

Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:163]: @cescoffier what would be the purpose of the stork build item? Mark that initialization is done?

File: extensions/resteasy-reactive/rest-client-reactive/runtime/pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/runtime/pom.xml:23]: will do, I need another relase of Stork for this. If you want to review the change in Stork: https://github.com/smallrye/smallrye-stork/pull/77 :)

File: integration-tests/rest-client-reactive-stork/src/test/java/io/quarkus/it/rest/reactive/stork/RestClientReactiveStorkTest.java (hunk 9 above)
Comment (michalszynkiewicz) [reply integration-tests/rest-client-reactive-stork/src/test/java/io/quarkus/it/rest/reactive/stork/RestClientReactiveStorkTest.java:26]: done

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java (hunk 11 above)
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java]: @geoand fixed

File: extensions/resteasy-reactive/rest-client-reactive/runtime/pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/runtime/pom.xml]: done

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java (hunk 11 above)
Comment (cescoffier) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java]: Still trying to see what would be possible. Typically, can we check at build time that the configured discovery and load balancing strategy exist.

Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java]: That would mean that service discovery type has to be build time fixed, no?
I'm not sure we want that

**Review by cescoffier — APPROVED** [review]
LGTM

**Comment by michalszynkiewicz** [comment]
the MP TCK failure seems unrelated, it fails on the fault tolerance TCK
